enhancements/kube-apiserver/stale-clusteroperators-detection.md

```diff
@@ -0,0 +1,87 @@
+---
+title: detection-and-reporting-of-stale-clusteroperators
+authors:
+  - "@mfojtik"
+reviewers:
+  - "@deads2k"
+approvers:
+  - "@deads2k"
+creation-date: 2022-01-31
+last-updated: 2022-01-31
+---
+
+# Detection and reporting of stale cluster operators
+
+## Summary
+
+In OpenShift 4, every active operator included in the core payload (cluster operator) updates its conditions in "
+clusteroperator" resource. Conditions determine whether specific core components are Degraded or Available or if they
+are making progress or if their state causes the cluster to be not upgradeable. These conditions are essential for
+cluster administrators to determine the health of their clusters, and they are usually the first thing Red Hat support
+is going to look at in case there is a problem with the cluster.
+
+Unfortunately, in some edge cases, these conditions might not reflect the current state of the cluster. For example,
+since they depend on the running operator, these conditions might become stale and outdated if the operator is not
+running or cannot contact the API server. And while the API server might still work, the result
+of `oc get clusteroperators` does not represent the cluster's current state.
+
+Therefore, we need a detection mechanism to help detect in-active operators or operators that are not actively updating
+the conditions.
+
+## Motivation
+
+In the past, there were multiple cases when either kubelet or CRI-O failed to run pods, either by KCM misfunctioning or
+some other system error. However, the static pods succeeded in running, making the Kubernetes API server available and
+accessible with `oc`. When support or engineers ran `oc get clusteroperators` against this cluster, they saw all
+operators reporting available, non-degraded and upgradeable status, but the truth was that none of the operators that
+provided these conditions were running.
+
+### Goals
+
+* Detect in-active cluster operators
+* Provide more accurate condition messages and status
+
+### Non-Goals
+
+* Make a new operator that detects other operators status
+* Create a heart-beat mechanism
+
+## Proposal
+
+To implement stale operator detection, we need a controller process that will run close to the Kubernetes API server,
+preferably using a local connection and client-based certificate for authentication to avoid dependencies on working
+auth and network when the cluster is unstable.
+
+The controller will watch the clusteroperator resource and check the `status.[]conditions.lastTransitionTime` field for
+Degraded, Available, Progressing, and Upgradeable conditions. Suppose the "lastTransitionTime" field is not updated for
+10 minutes. In that case, the controller will change the human-readable condition message
```

Comment:
`lastTransitionTime` is normally only [updated when the status transitions](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go#L43-L50) from once value to another. Futhermore, we probably won't want to change that because it is an important source of information - if the operator is up, this field tells you how long it has been up.

I assume that most operators will update the message again, but using the lastTransitionTime to remember state about when you last checked is problematic.

Reply:
> I assume that most operators will update the message again, but using the lastTransitionTime to remember state about when you last checked is problematic.

it's not perfect, but it's not problematic.  Since this a non-destructive write that does not change the condition.status, the human readable message tweak is sufficient when combined with an in-memory recording of, "I touched these".  That will limit the number of writes while providing the indications.

Reply:
I agree that an in-memory recording of the last checked times would work (it could even be made a little more robust by adding a timestamp to the message it sets). 

As currently written though, this implies that it will look at the `lastTransitionTime` and that won't work - if this new controller doesn't change the time, then the operator will set it back immediately to the original state and you will go immediately into an inifinite hot loop. And I don't think we want the new controller to modify the `lastTransitionTime` because that changes the meaning of the field.

Reply:
(the message will be reassert in short order, ideally, the lastTransitionTime will not be restored)

Reply:
> how is it non-destructive? it destroys the current lastTransitionTime and the current message associated w/ the condition

It should not.  Changing a message does not change the status and the lastTransitionTime is about status, not the message.  If an operator is reseting lastTransitionTime based on the message change, then that operator is bugged: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1497

Reply:
got it, this is resolved from my perspective, dunno if @zaneb still has concerns

Reply:
No concerns with what is being discussed in the comments here (store last checked time in-memory; change message at 10-minute intervals unless lastTransitionTime is <10 minutes ago). Slight concern that the actual text of the enhancement could be interpreted to mean something else.